KIP-0017: wallet-connect-v2 sign-api session spec (#31)
kadena-io/KIPs: kip-0017.md

        - [kadena_quickSign_v1 Response](#kadena_quicksign_v1-response)
      - [kadena_sign_v1](#kadena_sign_v1)
        - [Example kadena_sign_v1 Request and Response](#example-kadena_sign_v1-request-and-response)
        - [kadena_sign_v1 Request](#kadena_sign_v1-request)
        - [kadena_sign_v1 Response](#kadena_sign_v1-response)
    - [WalletConnect Events](#walletconnect-events)
- [Rationale for the spec](#rationale-for-the-spec)
  - [Accounts vs Public Keys](#accounts-vs-public-keys)
  - [SLIP-0044](#slip-0044)
- [Backwards Compatibility](#backwards-compatibility)

# Motivation

This KIP proposes that the Kadena ecosystem implement the
[Wallet Connect v2 Sign API](https://docs.walletconnect.com/2.0/api/sign). This
API allows for establishing a secure channel for communicating between dApps and
wallets in a platform-agnostic manner.

This is beneficial to Kadena for the following reasons:

- The Kadena ecosystem doesn't have a standard protocol for wallets and dApps to
  communicate with each other.
- The Wallet Connect (WC) protocol is tried and
  tested across all devices and it removes the burden of having to maintain a

>   kadena-specific implementation.

Reviewer: ```suggestion
  Kadena-specific implementation.
```

The response for the Proposal contains a `namespaces` object with the following properties:

1. `namespaces`: `Object`\
    1.1. `accounts`: `Array<string>` - an array of [Accounts](#wallet-connect-accounts) availble for signing.\
    1.2. `methods`: `Array<string>` - an array of [Methods](#walletconnect-methods) available.\
    1.3. `events`: `Array<string>` - an array of [Events](#walletconnect-events) available.

## Wallet Connect Chain IDs

Chain IDs are Wallet Connect’s way of identifying specific blockchains within an ecosystem.

They are defined by the [CAIP-2 standard](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) and are formatted as:

```
    <namespace> + ":" + <reference>
```
- `<namespace>`: `string` - always defined as "kadena".
- `<reference>`: `string` - a unique network ID within the blockchain ecosystem (e.g. `testnet04`, `mainnet01`).

These are valid Chain IDs:

- `kadena:mainnet01` - Kadena mainnet
- `kadena:testnet04` - Current Kadena testnet
- `kadena:development` - Kadena devnet

> **NB**: `Chain IDs`, as defined here, are semantically specific to
> WalletConnect and are not equivalent to a Kadena "Chain ID" (e.g. chain 1).

> > They are more closely related to Kadena's `networkId` field.

Reviewer: Why aren't they called "Network IDs", then?

Reviewer 2: In this context the name "Chain ID" is defined by the CAIP-2 standard which the Wallet Connect specification adheres to. We're unable to change that name, but merely added this note to avoid confusion with Kadena's Chain ID.

Reviewer 3: We want to use the naming that WalletConnect uses in places where it's dictated by WalletConnect

> We are displaying Kadena `chainId` information in the [`kadena_getAccounts_v1`](#kadena_getAccounts_v1) method.

## Wallet Connect Accounts

> Accounts are Wallet Connect's way of specifying control over signing priviledges.

Reviewer: ```suggestion
Accounts are Wallet Connect's way of specifying control over signing privileges.
```

They are defined by the [CAIP-10 standard](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md), and are formatted as:
```
    <chain_id> + ":" + <account_address>
```
- `<chain_id>`: `string` - the Wallet Connect [Chain ID](#wallet-connect-chain-ids).
- `<account_address>`: `string` - the public key available for signing by the wallet.

```jsonc
// Wallet Connect Account Example

"kadena:mainnet01:38298612cc2d5e841a232bd08413aa5304f9ef3251575ee182345abc3807dd89"
```

> **NB**: `accounts` is semantically specific to WalletConnect. This does not
> represent an account in a Kadena fungible contract (e.g. `coin`).
> Read the motivation for this in [Accounts vs. Public Keys](#accounts-vs-public-keys).

> Some users will want to know which account names in a Kadena fungible contract are associated with the public keys returned by Wallet Connect.
> 
> These users have two options:
> - Some account names can be derived from the public key (e.g. `k:` accounts that are defined as `k:<public_key>`).
> - For all other cases, the [`kadena_getAccounts_v1`](#kadena_getAccounts_v1)
> method can be called.

Reviewer 4: Adding "Kadena" in front of accounts for some clarity.

```suggestion
After the initial pairing with the wallet via WalletConnect, the Kadena accounts relevant for each public key can be requested by calling [getAccounts](#kadena_getaccounts_v1).

The use of account names that can be derived from the public key (i.e. `k:`-accounts that are defined as `k:<public_key>`), is **highly discouraged** as implementing this will limit the end-users flexibility in which account to use, thus making dApps less generic.
```

## WalletConnect Methods

Wallet Connect methods are ways for dApps to interact with a wallet using predefined method identifiers and an optional payload.

The methods available for Kadena are:

- [`kadena_getAccounts_v1`](#kadena_getAccounts_v1)

> - [`kadena_quickSign_v1`](#kadena_quicksign_v1)

Reviewer 4: Yes, there are a couple of places that we use the `kadena_quickSign_v` string literal, so we have to change it in all those places.

- [`kadena_sign_v1`](#kadena_sign_v1)

Any additional Kadena method names will be formatted as:
```

>     kadena_<methodName>_v<methodVersionNumber>

Reviewer 4: There are other places that add this tab

```
- `<methodName>`: `string` - name of the method in camelCase format.
- `<methodVersionNumber>`: `integer` - the version number of the method.

> **NB**: All method names will be versioned to allow for future
> upgradability of the Wallet Connect method APIs.

### kadena_getAccounts_v1

This method returns the fungible account names associated with a specific Wallet Connect account (i.e. a prefixed public key) in the Kadena blockchain.

These Wallet Connect accounts are retrieved from the `accounts` property of a [Paring response](#example-pairing-proposal-request-and-settlement-response).

> 

Reviewer 4: The `(i.e. a prefixed public key)` was an attempt to quickly describe what we are returning in WalletConnect accounts. WalletConnect accounts in the Kadena implementation are public keys with a prefix (`kadena:testnet04:` for example). 

In this section, there are two types of accounts we are using, kadena account and WalletConnect accounts, so I wanted to add some additional descriptions.

> 

Reviewer 3: Ah yes, maybe an example would be more describing?

> 

Reviewer 4: I edited the suggestion to use `Kadena account name` to help clarify which account we're referring to.

This method expects a `method` field that is set to `"kadena_sign_v1"`.

This method also expects a `params` object with the following properties:

1. `params`: `Object` - A [`SigningCmdRequest`](https://kadena-io.github.io/signing-api/#/definitions/SigningRequest) containing information the wallet needs to create and sign the transaction.

#### kadena_sign_v1 Response

The method response expects a `result` object with the following properties:

1. `result`: `Object` - A [`SigningResponse`](https://kadena-io.github.io/signing-api/#/definitions/SigningResponse) that contains a signed command ready to be submitted to the blockchain.

## WalletConnect Events

At the moment we see no need to standardize any events and believe that most
use cases can be handled at the Kadena protocol level that sits atop Wallet
Connect.

This is subject to change in the future as we understand the workflows more through extended use of the protocol.

However, any future Kadena event names will be formatted as:
```
    kadena_<eventName>_v<eventVersionNumber>
```
- `<eventName>`: `string` - name of the event in camelCase format.
- `<eventVersionNumber>`: `integer` - the version number of the event.

> > **NB**: All event names will be versioned to allow for facilitate future

Reviewer: ```suggestion
> **NB**: All event names will be versioned to allow for future
```

> upgradability of the Wallet Connect events APIs.

# Rationale for the spec

## Accounts vs Public Keys

Several options were considered when discussing what information to include
in the `accounts` section of the initial pairing response.

However, the following reasons forced us to only include the public keys
(prefixed with Kadena's Wallet Connect Chain ID):

1. Kadena fungible contracts (specifically the `coin` contract) have a maximum account name length of 256, but
   the [CAIP-10 standard](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md)
   limits wallet connect account names to a maximum of 128 characters.

> 2. We considered only supporting `k:` accounts, as they're build from

Reviewer: ```suggestion
2. We considered only supporting `k:` accounts, as they're built from
```

   `k`:`publicKey`, but we want this response to be generic for all accounts
   that could exist on the Kadena blockchain.

> Therefore, we **strongly** advice builders to integrate the [`kadena_getAccounts_v1`](#kadena_getAccounts_v1)

Reviewer: ```suggestion
Therefore, we **strongly** advise builders to integrate the [`kadena_getAccounts_v1`](#kadena_getAccounts_v1)
```

method into the pairing process with the wallet.

This will allow them to support multi-sig accounts (e.g. `w:`-accounts) and

> other means of singing transactions.

Reviewer: ```suggestion
other means of signing transactions.
```

We do **NOT** recommend extracting the `publicKey` from the Wallet Connect account
and just adding a `k:` in front of it.

## SLIP-0044

While not required in the namespace proposal process, a network's SLIP-0044
value is used throughout the WalletConnect demo code.

Kadena's coin type is
**626** as documented in

> https://github.com/satoshilabs/slips/blob/master/slip-0044.md.

Reviewer: ```suggestion
(SLIP-0044)[https://github.com/satoshilabs/slips/blob/master/slip-0044.md].
```
